lib/OpenQA/WebAPI/Controller/Step.pm

```diff
@@ -91,6 +118,63 @@ sub view ($self) {
     $self->viewimg;
 }
 
+=head2 _determine_needles_dir_for_job
+
+my ($needle_dirs, $needles_var) = $self->_determine_needles_dir_for_job($job);
+
+Get the directory where needle files are stored for a given job.  
+It checks whether the Git-based needle management is enabled.
+If enabled, it fetches the path to the needles directory and the job setting NEEDLES_DIR (or CASEDIR as fallback).
+
+=head3 Parameters
+
+=over 4
+
+=item * $job - OpenQA job
+
+=back
+
+=head3 Returns
+
+=over 4
+
+=item * A list containing:
+
+    - The absolute path to the needles directory.
+    - The job settings result object corresponding to the needles directory (either `NEEDLES_DIR` or `CASEDIR`).
+
+=back
+
+=cut
+sub _determine_needles_dir_for_job ($self, $job) {
+    return undef unless $self->app->config->{'scm git'}->{checkout_needles_sha} eq 'yes';
+    return undef unless my $needle_dirs = realpath($job->needle_dir);
+    my $settings = $job->settings;
+    return ($needle_dirs, $settings->single({key => 'NEEDLES_DIR'}) // $settings->single({key => 'CASEDIR'}));
+}
+
+sub _create_tmpdir_for_needles_refspec ($self, $job) {
+    my ($needle_dirs, $needles_dir_var) = $self->_determine_needles_dir_for_job($job);
+    return undef unless $needles_dir_var;
+    my $needles_url = Mojo::URL->new($needles_dir_var->value);
+    return undef unless $needles_url->scheme;
+    my $needles_ref = $needles_url->fragment;
+    eval {
+        my $vars = decode_json(path($job->result_dir, 'vars.json')->slurp);
+        $needles_ref = $vars->{NEEDLES_GIT_HASH} if ref $vars eq 'HASH';
+    };
```

Comment:
That eval is never checked. The question is, why is it wrapped in an eval? Which part is ok to fail here?
Should it accept a broken `var.json`? Or a non existing one?
I think an eval around that whole part is not a good design.

The thing is, the test below creates the `vars.json` in the wrong directory, so the file here doesn't exist. The test passes because it gets the sha from git git call later.

Reply:
I found it there before I started working on it so I figured there was a reason for, also while I was doing some tests I discover that if the var.json not exists it rightly throw errors.

Reply:
I've removed the eval section and put a check for undef value for job->result_dir

Reply:
This should also let the coverage be happy even if we have test if the job->result_dir is not empty

Reply:
I wrote:
> The thing is, the test below creates the `vars.json` in the wrong directory, so the file here doesn't exist. The test passes because it gets the sha from git git call later.

You are still creating the vars.json in the wrong directory. That's not the directory where this file should end up.
And both things should be tested.
The fact that every line is covered now because of `... if $x` does not mean that the statements are actually covered.

We should make sure that the part with the `NEEDLES_GIT_HASH` is working as well.